File: cli-validation/python-cli/src/documentation_robotics/core/dependency_tracker.py
"""
Dependency tracker - analyzes dependencies between elements.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Any, Dict, List, Optional, Set

import networkx as nx

from .reference_registry import ReferenceRegistry


class TraceDirection(Enum):
    """Direction to trace dependencies."""

    UP = "up"  # Find what this depends on
    DOWN = "down"  # Find what depends on this
    BOTH = "both"  # Both directions


@dataclass
class DependencyPath:
    """Represents a dependency path between two elements."""

    source: str
    target: str
    path: List[str]
    depth: int
    relationship_types: List[str]


class DependencyTracker:
    """
    Tracks and analyzes dependencies between elements.

    Uses reference registry and graph algorithms to provide
    dependency analysis and impact assessment.
    """

    def __init__(self, model: Any):
        """
        Initialize dependency tracker.

        Args:
            model: The architecture model
        """
        self.model = model
        self.registry = ReferenceRegistry()
        self._build_registry()

    def _build_registry(self) -> None:
        """Build reference registry from model."""
        for layer in self.model.layers.values():
            for element in layer.elements.values():
                self.registry.register_element(element)

    def trace_dependencies(
        self,
        element_id: str,
        direction: TraceDirection = TraceDirection.BOTH,
        max_depth: Optional[int] = None,
    ) -> List[Any]:
        """
        Trace dependencies from an element.

        Args:
            element_id: Element ID to start from
            direction: Direction to trace (up, down, both)
            max_depth: Maximum depth to trace

        Returns:
            List of dependent elements
        """
        graph = self.registry.get_dependency_graph()

        if element_id not in graph:
            return []

        dependencies = set()

        # Trace upward (what this element depends on)
        if direction in [TraceDirection.UP, TraceDirection.BOTH]:
            dependencies.update(self._trace_up(graph, element_id, max_depth))

        # Trace downward (what depends on this element)
        if direction in [TraceDirection.DOWN, TraceDirection.BOTH]:
            dependencies.update(self._trace_down(graph, element_id, max_depth))

        # Convert IDs to Elements
        elements = []
        for dep_id in dependencies:
            element = self.model.get_element(dep_id)
            if element:
                elements.append(element)

        return elements

    def _trace_up(self, graph: nx.DiGraph, element_id: str, max_depth: Optional[int]) -> Set[str]:
        """Trace upward dependencies (what element depends on)."""
        if max_depth is None:
            # Get all descendants (what this points to / depends on)
            return nx.descendants(graph, element_id)
        else:
            # Get descendants within depth
            descendants = set()
            current_level = {element_id}

            for _ in range(max_depth):
                next_level = set()
                for node in current_level:
                    if node in graph:
                        next_level.update(graph.successors(node))
                descendants.update(next_level)
                current_level = next_level
                if not current_level:
                    break

            return descendants

    def _trace_down(self, graph: nx.DiGraph, element_id: str, max_depth: Optional[int]) -> Set[str]:
        """Trace downward dependencies (what depends on element)."""
        if max_depth is None:
            # Get all ancestors (what points to this / depends on this)
            return nx.ancestors(graph, element_id)
        else:
            # Get ancestors within depth
            ancestors = set()
            current_level = {element_id}

            for _ in range(max_depth):
                next_level = set()
                for node in current_level:
                    if node in graph:
                        next_level.update(graph.predecessors(node))
                ancestors.update(next_level)
                current_level = next_level
                if not current_level:
                    break

            return ancestors

    def find_dependency_paths(
        self, source_id: str, target_id: str, max_paths: int = 10
    ) -> List[DependencyPath]:
        """
        Find all dependency paths between two elements.

        Args:
            source_id: Source element ID
            target_id: Target element ID
            max_paths: Maximum number of paths to return

        Returns:
            List of dependency paths
        """
        graph = self.registry.get_dependency_graph()

        if source_id not in graph or target_id not in graph:
            return []

        try:
            # Find all simple paths
            paths = list(nx.all_simple_paths(graph, source_id, target_id, cutoff=max_paths))

            # Convert to DependencyPath objects
            dependency_paths = []
            for path in paths[:max_paths]:
                # Get relationship types along path
                rel_types = []
                for i in range(len(path) - 1):
                    edge_data = graph.get_edge_data(path[i], path[i + 1])
                    rel_types.append(edge_data.get("type", "unknown") if edge_data else "unknown")

                dependency_paths.append(
                    DependencyPath(
                        source=source_id,
                        target=target_id,
                        path=path,
                        depth=len(path) - 1,
                        relationship_types=rel_types,
                    )
                )

            return dependency_paths

        except nx.NetworkXNoPath:
            return []

    def get_dependency_layers(self, element_id: str) -> Dict[str, List[str]]:
        """
        Get dependencies organized by layer.

        Args:
            element_id: Element ID

        Returns:
            Dictionary mapping layer name to list of dependent element IDs
        """
        dependencies = self.trace_dependencies(element_id, TraceDirection.BOTH)

        by_layer: Dict[str, List[str]] = {}

        for dep in dependencies:
            if dep.layer not in by_layer:
                by_layer[dep.layer] = []
            by_layer[dep.layer].append(dep.id)

        return by_layer

    def get_orphaned_elements(self) -> List[Any]:
        """
        Find elements with no references (incoming or outgoing).

        Returns:
            List of orphaned elements
        """
        graph = self.registry.get_dependency_graph()
        orphaned = []

        for layer in self.model.layers.values():
            for element in layer.elements.values():
                if element.id not in graph:
                    orphaned.append(element)
                elif graph.degree(element.id) == 0:
                    orphaned.append(element)

        return orphaned

    def get_hub_elements(self, threshold: int = 10) -> List[tuple]:
        """
        Find hub elements (elements with many connections).

        Args:
            threshold: Minimum number of connections to be considered a hub

        Returns:
            List of (element_id, connection_count) tuples
        """
        graph = self.registry.get_dependency_graph()
        hubs = []

        for node in graph.nodes():
            degree = graph.degree(node)
            if degree >= threshold:
                hubs.append((node, degree))

        # Sort by degree descending
        hubs.sort(key=lambda x: x[1], reverse=True)

        return hubs

File: cli/src/core/dependency-tracker.ts
import Graph from 'graphology';
import { ReferenceRegistry } from './reference-registry.js';

/**
 * Direction to trace dependencies (matches Python CLI TraceDirection enum)
 */
export enum TraceDirection {
  UP = 'up',     // Find what this depends on (successors/descendants)
  DOWN = 'down', // Find what depends on this (predecessors/ancestors)
  BOTH = 'both'  // Both directions
}

/**
 * Represents a dependency path between two elements
 */
export interface DependencyPath {
  source: string;
  target: string;
  path: string[];
  depth: number;
  relationship_types: string[];
}

/**
 * Dependency tracker - analyzes dependencies between elements
 *
 * Uses reference registry and graph algorithms to provide
 * dependency analysis and impact assessment.
 *
 * Matches Python CLI behavior from dependency_tracker.py
 */
export class DependencyTracker {
  private registry: ReferenceRegistry;

  constructor(registry: ReferenceRegistry) {
    this.registry = registry;
  }

  /**
   * Trace dependencies from an element
   *
   * @param elementId - Element ID to start from
   * @param direction - Direction to trace (up, down, both)
   * @param maxDepth - Maximum depth to trace (null = unlimited)
   * @returns Array of dependent element IDs
   */
  traceDependencies(
    elementId: string,
    direction: TraceDirection = TraceDirection.BOTH,
    maxDepth: number | null = null
  ): string[] {
    const graph = this.registry.getDependencyGraph();

    if (!graph.hasNode(elementId)) {
      return [];
    }

    const dependencies = new Set<string>();

    // Trace upward (what this element depends on - successors)
    if (direction === TraceDirection.UP || direction === TraceDirection.BOTH) {
      const upDeps = this._traceUp(graph, elementId, maxDepth);
      upDeps.forEach(id => dependencies.add(id));
    }

    // Trace downward (what depends on this element - predecessors)
    if (direction === TraceDirection.DOWN || direction === TraceDirection.BOTH) {
      const downDeps = this._traceDown(graph, elementId, maxDepth);
      downDeps.forEach(id => dependencies.add(id));
    }

    return Array.from(dependencies);
  }

  /**
   * Trace upward dependencies (what element depends on - successors)
   * Private helper matching Python CLI behavior
   */
  private _traceUp(graph: Graph, elementId: string, maxDepth: number | null): Set<string> {
    if (maxDepth === null) {
      // Get all descendants (unlimited depth)
      return this._getDescendants(graph, elementId);
    } else {
      // Get descendants within depth using BFS
      return this._getDescendantsLimited(graph, elementId, maxDepth);
    }
  }

  /**
   * Trace downward dependencies (what depends on element - predecessors)
   * Private helper matching Python CLI behavior
   */
  private _traceDown(graph: Graph, elementId: string, maxDepth: number | null): Set<string> {
    if (maxDepth === null) {
      // Get all ancestors (unlimited depth)
      return this._getAncestors(graph, elementId);
    } else {
      // Get ancestors within depth using BFS
      return this._getAncestorsLimited(graph, elementId, maxDepth);
    }
  }

  /**
   * Get all descendants (reachable via outgoing edges) - DFS
   * Equivalent to networkx.descendants()
   */
  private _getDescendants(graph: Graph, startNode: string): Set<string> {
    const visited = new Set<string>();
    const stack = [startNode];

    while (stack.length > 0) {
      const node = stack.pop()!;

      if (!graph.hasNode(node)) continue;

      for (const neighbor of graph.outNeighbors(node)) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          stack.push(neighbor);
        }
      }
    }

    return visited;
  }

  /**
   * Get ancestors (reachable via incoming edges) - DFS
   * Equivalent to networkx.ancestors()
   */
  private _getAncestors(graph: Graph, startNode: string): Set<string> {
    const visited = new Set<string>();
    const stack = [startNode];

    while (stack.length > 0) {
      const node = stack.pop()!;

      if (!graph.hasNode(node)) continue;

      for (const neighbor of graph.inNeighbors(node)) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          stack.push(neighbor);
        }
      }
    }

    return visited;
  }

  /**
   * Get descendants within max_depth using BFS level-by-level
   * Matches Python CLI behavior exactly
   */
  private _getDescendantsLimited(graph: Graph, startNode: string, maxDepth: number): Set<string> {
    const descendants = new Set<string>();
    let currentLevel = new Set([startNode]);

    for (let depth = 0; depth < maxDepth; depth++) {
      const nextLevel = new Set<string>();

      for (const node of currentLevel) {
        if (!graph.hasNode(node)) continue;

        for (const neighbor of graph.outNeighbors(node)) {
          nextLevel.add(neighbor);
        }
      }

      nextLevel.forEach(id => descendants.add(id));
      currentLevel = nextLevel;

      if (currentLevel.size === 0) {
        break;
      }
    }

    return descendants;
  }

  /**
   * Get ancestors within max_depth using BFS level-by-level
   * Matches Python CLI behavior exactly
   */
  private _getAncestorsLimited(graph: Graph, startNode: string, maxDepth: number): Set<string> {
    const ancestors = new Set<string>();
    let currentLevel = new Set([startNode]);

    for (let depth = 0; depth < maxDepth; depth++) {
      const nextLevel = new Set<string>();

      for (const node of currentLevel) {
        if (!graph.hasNode(node)) continue;

        for (const neighbor of graph.inNeighbors(node)) {
          nextLevel.add(neighbor);
        }
      }

      nextLevel.forEach(id => ancestors.add(id));
      currentLevel = nextLevel;

      if (currentLevel.size === 0) {
        break;
      }
    }

    return ancestors;
  }

  /**
   * Find all dependency paths between two elements
   *
   * @param sourceId - Source element ID
   * @param targetId - Target element ID
   * @param maxPaths - Maximum number of paths to return
   * @returns Array of dependency paths
   */
  findDependencyPaths(
    sourceId: string,
    targetId: string,
    maxPaths: number = 10
  ): DependencyPath[] {
    const graph = this.registry.getDependencyGraph();

    if (!graph.hasNode(sourceId) || !graph.hasNode(targetId)) {
      return [];
    }

    // Find all simple paths (no cycles)
    const paths = this._findAllSimplePaths(graph, sourceId, targetId, maxPaths);

    // Convert to DependencyPath objects
    return paths.map(path => {
      // Get relationship types along path
      const relationshipTypes: string[] = [];
      for (let i = 0; i < path.length - 1; i++) {
        const edgeAttrs = graph.getEdgeAttributes(path[i], path[i + 1]);
        relationshipTypes.push(edgeAttrs?.type || 'unknown');
      }

      return {
        source: sourceId,
        target: targetId,
        path,
        depth: path.length - 1,
        relationship_types: relationshipTypes
      };
    });
  }

  /**
   * Find all simple paths between source and target
   * Implements DFS path-finding similar to networkx.all_simple_paths
   */
  private _findAllSimplePaths(
    graph: Graph,
    source: string,
    target: string,
    maxPaths: number
  ): string[][] {
    const paths: string[][] = [];
    const visited = new Set<string>();

    const dfs = (current: string, path: string[]) => {
      if (paths.length >= maxPaths) {
        return;
      }

      if (current === target) {
        paths.push([...path, current]);
        return;
      }

      visited.add(current);
      path.push(current);

      if (graph.hasNode(current)) {
        for (const neighbor of graph.outNeighbors(current)) {
          if (!visited.has(neighbor)) {
            dfs(neighbor, path);
          }
        }
      }

      path.pop();
      visited.delete(current);
    };

    dfs(source, []);
    return paths;
  }

  /**
   * Find hub elements (elements with many connections)
   *
   * @param threshold - Minimum number of connections to be considered a hub
   * @returns Array of [elementId, connectionCount] tuples, sorted by count descending
   */
  getHubElements(threshold: number = 10): [string, number][] {
    const graph = this.registry.getDependencyGraph();
    const hubs: [string, number][] = [];

    for (const node of graph.nodes()) {
      const degree = graph.degree(node);
      if (degree >= threshold) {
        hubs.push([node, degree]);
      }
    }

    // Sort by degree descending
    hubs.sort((a, b) => b[1] - a[1]);

    return hubs;
  }
}
